### PR TITLE
Handle cluster chain switchover at epoch boundary

### PR DESCRIPTION
**Pull request by jordanschalm**
Issue: https://github.com/dapperlabs/flow-go/issues/4480

This PR begins implementation of the process of transitioning to a new epoch for the collection node. When we transition to a new epoch, we spin up a hotstuff instance, block sync engine, and compliance/proposal engine for that epoch. When we can no longer produce valid collections for the previous epoch (see [details](https://www.notion.so/dapperlabs/Transaction-Routing-e4a46d1f24a44985ae12d0495ab530c0)), we shut down the epoch-scoped components for the previous epoch.

### Changes
* Update `epochmgr` engine to handle epoch transition events
* Add `events.Heights` "gadget", which enables setting up callbacks for particular block heights
* Add utilities for concurrently starting/stopping multiple `ReadyDone` instances
* Add `EpochQuery` mock for simplifying working with epochs in tests

### CODE REVIEW

File: engine/collection/epochmgr/engine.go [hunk 1]
```diff
@@ -108,25 +139,97 @@ func (e *Engine) createEpochComponents(epoch protocol.Epoch) (*EpochComponents,
...
+	// start up components
+	err = e.startEpochComponents(counter, components)
+	if err != nil {
+		return fmt.Errorf("could not start epoch components: %w", err)
+	}
+
+	log.Info().Msg("epoch transition: new epoch components started successfully")
+
+	// set up trigger to shut down previous epoch components after max expiry
+	e.heightEvents.OnHeight(first.Height+flow.DefaultTransactionExpiry, func() {
```
Comment (zhangchiqing) [inline engine/collection/epochmgr/engine.go:208]: This is a tricky problem. I'm not sure if the solution here would work well. This reminds me [the issue I ran into when implementing consensus tests](https://github.com/onflow/flow-go/blob/b6798cc07caac3809fe5b0d637a3ab54399ffa2d/consensus/integration/stopper_test.go#L27-L33). The issue is the faster nodes who have reached the targeted height will stop hotstuff, which will cause other slow nodes unable to get the blocks to finalize their targeted height, and those slow nodes will stuck forever.

In the consensus tests, I solve the issue by [letting faster nodes keep building blocks until all the honest nodes have all finalized enough blocks, and then stop them all](https://github.com/onflow/flow-go/blob/b6798cc07caac3809fe5b0d637a3ab54399ffa2d/consensus/integration/stopper_test.go#L55-L71). 

I feel we might want something similar here.

File: engine/collection/epochmgr/engine.go [hunk 2]
```diff
@@ -87,17 +114,21 @@ func (e *Engine) Ready() <-chan struct{} {
...
-		<-e.epoch.hotstuff.Done()
-		<-e.epoch.prop.Done()
-		<-e.epoch.sync.Done()
+		// Stop components for all epochs. This is typically a single epoch
+		// but can be multiple near epoch boundaries
+		epochs := make([]module.ReadyDoneAware, 0, len(e.epochs))
+		for _, epoch := range e.epochs {
+			epochs = append(epochs, epoch)
+		}
+		<-lifecycle.AllDone(epochs...)
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:131]: `e.epochs` is a map, not a slice

File: engine/collection/epochmgr/engine.go (hunk 1 above)
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:208]: I believe what you're describing doesn't apply here. Let me know if I'm misunderstanding.

We're dealing with two different chains here: the cluster chain and the main chain. We're triggering the end of the cluster chain based on reaching a particular height on the main chain. Progress on the main chain will continue independently of individual collectors shutting down their cluster chain, so all collectors will eventually shut down their cluster chain so long as they continue following the main chain.

Comment (zhangchiqing) [reply engine/collection/epochmgr/engine.go:208]: Sorry, I think you are right :) 

Glad this is not an issue.

Comment (zhangchiqing) [reply engine/collection/epochmgr/engine.go:208]: I would suggest to add some explanation on why the timing is chosen

**Review by zhangchiqing — APPROVED** [review]
Nice work!

File: engine/collection/epochmgr/engine.go (hunk 1 above)
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:208]: > Glad this is not an issue.

Me too 😀 -- I'll add some docs to explain what's going on here in more detail

File: cmd/collection/main.go [hunk 3]
```diff
@@ -339,12 +340,16 @@ func main() {
 				syncFactory,
 			)
 
+			heightEvents := gadgets.NewHeights()
+			node.ProtocolEvents.AddConsumer(heightEvents)
```
Comment (vishalchangrani) [inline cmd/collection/main.go:344]: is this what I should be subscribe to as well to be notified about epochs at the networking layer?

Comment (jordanschalm) [reply cmd/collection/main.go:344]: Yes, `node.ProtocolEvents` is what you want to use 👍

Comment (vishalchangrani) [reply cmd/collection/main.go:344]: sweet

**Review by vishalchangrani — APPROVED** [review]
lgtm!

File: engine/collection/epochmgr/engine.go (hunk 1 above)
Comment (jordanschalm) [reply engine/collection/epochmgr/engine.go:215]: Added a note in https://github.com/onflow/flow-go/pull/2/commits/2b39f85451857ad343940b84db8f6c550af91d02

**Pull request merged**